scheduler/reconcile.go

```diff
@@ -1045,95 +1054,152 @@ func (a *allocReconciler) computeStop(group *structs.TaskGroup, nameIndex *alloc
-			a.result.stop = append(a.result.stop, allocStopResult{
-				alloc:             stopAlloc,
-				statusDescription: statusDescription,
-			})
-			delete(deleteSet, stopAlloc.ID)
+			// Pick which allocation we want to keep.
+			keepAlloc := pickReconnectingAlloc(reconnectingAlloc, replacementAlloc)
+			if keepAlloc == replacementAlloc {
+				// The replacement allocation is preferred, so stop the one
+				// reconnecting if not stopped yet.
+				if _, ok := stop[reconnectingAlloc.ID]; !ok {
+					stop[reconnectingAlloc.ID] = reconnectingAlloc
+					a.result.stop = append(a.result.stop, allocStopResult{
+						alloc:             reconnectingAlloc,
+						statusDescription: allocNotNeeded,
+					})
+				}
 
-			remove--
-			// if we've removed all we need to, stop iterating and return.
-			if remove == 0 {
-				return remove
+				// Our assumption that all reconnecting allocations will
+				// reconnect was wrong, so remove this one from the reconnect
+				// set.
+				delete(reconnect, reconnectingAlloc.ID)
+			} else {
+				// The reconnecting allocation is preferred, so stop this
+				// replacement.
+				stop[replacementAlloc.ID] = replacementAlloc
+				a.result.stop = append(a.result.stop, allocStopResult{
+					alloc:             replacementAlloc,
+					statusDescription: allocReconnected,
+				})
 			}
 		}
 	}
 
-	return remove
+	return reconnect, stop
+}
+
+// pickReconnectingAlloc returns the allocation to keep between the original
+// one that is reconnecting and one of its replacements.
+//
+// This function is not commutative, meaning that pickReconnectingAlloc(A, B)
+// is not the same as pickReconnectingAlloc(B, A). Preference is given to keep
+// the original allocation when possible.
+func pickReconnectingAlloc(original *structs.Allocation, replacement *structs.Allocation) *structs.Allocation {
+	// Check if the replacement is newer.
+	// Always prefer the replacement if true.
+	replacementIsNewer := replacement.Job.Version > original.Job.Version ||
+		replacement.Job.CreateIndex > original.Job.CreateIndex
+	if replacementIsNewer {
+		return replacement
+	}
+
+	// Check if the replacement has better placement score.
+	// If any of the scores is not available, only pick the replacement if
+	// itself does have scores.
+	originalMaxScoreMeta := original.Metrics.MaxNormScore()
+	replacementMaxScoreMeta := replacement.Metrics.MaxNormScore()
+
+	replacementHasBetterScore := originalMaxScoreMeta == nil && replacementMaxScoreMeta != nil ||
+		(originalMaxScoreMeta != nil && replacementMaxScoreMeta != nil &&
+			replacementMaxScoreMeta.NormScore > originalMaxScoreMeta.NormScore)
+
+	// Check if the replacement has better client status.
+	// Even with a better placement score make sure we don't replace a running
+	// allocation with one that is not.
+	replacementIsRunning := replacement.ClientStatus == structs.AllocClientStatusRunning
+	originalNotRunning := original.ClientStatus != structs.AllocClientStatusRunning
```

Comment:
>  I had to build a big ol' truth table to verify that for myself 

Yup, I had to do the same thing, which should've been a signal that this logic is confusing 😅 

But I looked into it again, and I'm not sure if we can move this `ClientStatus` check to the caller because it's an awkward conditional tie-breaker.

The way I approached this was using a process of elimination. There are three states we can safely stop the reconnecting allocation without needing to compare it with its replacements:

* desired status not `run` (so `stop` or `evict`): handled by `stopReconnecting`, but now I'm wondering if we even need to stop in this case as opposed to just skip these allocs 🤷 
* client status `failed`: handled by `filterByFailedReconnect`, but we probably don't need this anymore and just handle it in the main loop.

That leaves 5 possible client status for original alloc:

* `lost` and `unknown`: I don't think these are possible for a reconnecting alloc? I think `lost` is only used for groups without `max_client_disconnect` (which is by definition not the case for allocation that are reconnecting) and the allocation would not be in the `reconnecting` set if it was `unknown`, so we can probably ignore these.
* `complete`: probably a no-op? Or maybe we still need to check and stop replacements? But only if not batch? Not sure, this seems a bit undefined 😅 
* `pending` and `running`: we need to compare with its replacements. If the replacement is newer we keep it, done. If the replacement has better score we need to check the client status.

And this is where the awkwardness lives. Outside this function we can't make a decision before comparing them and we don't know if the replacement was picked because it had a higher version or better placement (well, we can check again, but that would be kind of redundant).

>  I'm not sure what happens to the replacement in that case -- does it just fall thru to the rest of the reconciler?

That's right. My assumption here is the idea of merging timelines. When the client reconnects we assume the replacement allocations were being handled by the reconciler like any other, so if we don't need to stop them just let the reconciler keep managing it as it has been.

Reply:
> client status `failed`: handled by `filterByFailedReconnect`, but we probably don't need this anymore and just handle it in the main loop.

The `filterByFailedReconnect` should pick up the `complete` case as well because that's terminal.

> `pending` and `running`: we need to compare with its replacements. If the replacement is newer we keep it, done. If the replacement has better score we need to check the client status.

I probably could have explained it better, but I think we only need to compare if the replacement is running. So in other words, the check we're doing here could be before we compare scores and then exit early.

Reply:
> The `filterByFailedReconnect` should pick up the `complete` case as well because that's terminal.

It only looks for not _server_ terminal and `failed` explicitly, so I think `complete` is not handled.
https://github.com/hashicorp/nomad/blob/9a2fdb5f53dce81edf2802f0b64962e07596fd03/scheduler/reconcile_util.go#L529

> I think we only need to compare if the replacement is running

_...and has the same job version._ 

A new job version always wins because we don't want to keep outdated allocs, even if they are the original ones. So the client status check can be done before the score, but it can't be done before the job version check.

Reply:
Ah, good catch. Yeah it seems like there's no way to hoist this up.